### ISSUE TITLE
PHP 8.0 | PHP/NoSilencedErrors: review function list

### ISSUE THREAD

**Issue by jrfnl** [WordPress/WordPress-Coding-Standards#1971]
PHP 8 changes the behaviour of the error silence operator @ : where previously it would silence all notices, warnings and errors, it will now only silence notices and warnings, but will not prevent fatal errors/catchable exceptions anymore.

PHP 8 also contains another relevant change, which is that the error level for a number of notices/warnings has been [elevated to fatal error/catchable exception](https://wiki.php.net/rfc/engine_warnings#proposed_classification).

It would be very helpful if someone would review if any of the functions in the `$function_whitelist` are affected by these changes and need to be removed from the list.

Note: yes, the name of the property should also be changed, but there's a separate open ticket for that.

This was also discussed in the #core-php Slack channel: https://wordpress.slack.com/archives/C60K3MP2Q/p1611678080004700

**Comment by d-alleyne** [comment]
Hi @jrfnl 👋🏽 . 
I have started investigating this issue by setting up a test environment and reviewing the relevant code, particularly the [$allowedFunctionsList](https://github.com/WordPress/WordPress-Coding-Standards/blob/7722c0b8a6f4eb615be4ba8ebe22ed92a5fa87d7/WordPress/Sniffs/PHP/NoSilencedErrorsSniff.php#L86) in `NoSilencedErrorsSniff.php`. 

I notice that in the following file [NoSilencedErrorsUnitTest.inc](https://github.com/WordPress/WordPress-Coding-Standards/blob/develop/WordPress/Tests/PHP/NoSilencedErrorsUnitTest.inc), there are tests for a few I/O functions, e.g. `file_get_contents`, `file_exists`. Am I understanding it correctly that the goal is to add the missing functions of the `$allowedFunctionsList` array to this file, and compare their behaviour when executing under php7 and php8, and note any differences?

**Comment by jrfnl** [comment]
Hi @d-alleyne Lovely to hear that you are looking into this.

> Am I understanding it correctly that the goal is to add the missing functions of the `$allowedFunctionsList` array to this file, and compare their behaviour when executing under php7 and php8, and note any differences?

No, the goal is not to _add_ functions to the list, but to _remove_ them (where relevant).

As things were in PHP 5/7, a number of functions existed in PHP where no amount of error checking ahead of running the function would prevent a PHP notice/warning.
Those functions are the functions listed in the `$allowedFunctionsList`, as for those, it could be justified that the silence `@` operator was being used.

However, as of PHP 8.0, a number of PHP notices/warnings have been elevated to fatal errors and/or catchable exceptions.
If this has happened to the notice/warning for a function listed in the `$allowedFunctionsList`, the silence operator is now useless (when running code on PHP 8.0 or higher) and should not be used anymore.

Depending on the specific PHP 8.0 change affecting the function, the function call should be wrapped in a `try-`catch` or handled in a different way.

So basically, all functions in that list need to be checked to see if they are affected by the higher error levels and if so, they should likely be removed from the list to prevent giving people a false sense of security _"oh, I've silenced the warning"_, when the code on which the silence operator is being used will no longer have a silenced notice on PHP 8.0+.

Does that make it clearer ?

**Comment by jrfnl** [comment]
P.s.: as you mentioned setting up a test environment... in case you are not aware, http://3v4l.org/ with the "+ include eol" option selected in the dropdown is an awesome way to test this kind of behavioral change across different PHP versions.

**Comment by d-alleyne** [comment]
@jrfnl Do you mean something like [this](https://3v4l.org/7GHnS), but documented for all functions in `$allowedFunctionsList`?

**Comment by jrfnl** [comment]
@d-alleyne Not sure what you were trying to demonstrate with the code sample - here is a simplified version which shows the behaviour has not changed for that function: https://3v4l.org/l3jFs#veol

**Comment by d-alleyne** [comment]
@jrfnl My plan is to run these functions in both on PHP7 and PHP8, and document any differences, if any. In our existing example, it'd look like this.
e.g.
| Function | PHP7 | PHP8 |
|--------|--------|--------|
| [chdir](https://www.php.net/manual/en/function.chdir.php) | [Same Behaviour](https://3v4l.org/l3jFs#veol) | [Same Behaviour](https://3v4l.org/l3jFs#veol) |

**Comment by jrfnl** [comment]
@d-alleyne Sounds good. To document, I'm not sure the PHP 7/8 columns are needed as the code sample can be run against both in one go with 3v4l (using the "+ include eol" option selected in the dropdown). You can make links go straight to the results including EOL PHP versions by adding `#veol` to the link.

**Comment by d-alleyne** [comment]
Great idea, will do!

**Comment by d-alleyne** [comment]
| Function | Link | Remarks |
|----------|------|---------|
| [chdir](https://www.php.net/manual/en/function.chdir.php) | [https://3v4l.org/l3jFs#veol](https://3v4l.org/l3jFs#veol) | Behaviour unchanged |
| [opendir](https://www.php.net/manual/en/function.opendir.php) | [https://3v4l.org/DoMVG#veol](https://3v4l.org/DoMVG#veol) | Behaviour unchanged |
| [scandir](https://www.php.net/manual/en/function.scandir.php) | [https://3v4l.org/vf62L#veol](https://3v4l.org/vf62L#veol) | Behaviour unchanged |
| [file_exists](https://www.php.net/manual/en/function.file-exists.php) | [https://3v4l.org/G8C5L#veol](https://3v4l.org/G8C5L#veol) | Behaviour unchanged |
| [file_get_contents](https://www.php.net/manual/en/function.file-get-contents.php) | [https://3v4l.org/b0CaJ#veol](https://3v4l.org/b0CaJ#veol) | Behaviour unchanged |
| [file](https://www.php.net/manual/en/function.file.php) | [https://3v4l.org/9EpXo#veol](https://3v4l.org/9EpXo#veol) | Behaviour unchanged |
| [fileatime](https://www.php.net/manual/en/function.fileatime.php) | [https://3v4l.org/V6LJ6#veol](https://3v4l.org/V6LJ6#veol) | Behaviour unchanged |
| [filectime](https://www.php.net/manual/en/function.filectime.php) | [https://3v4l.org/RVIJG#veol](https://3v4l.org/RVIJG#veol) | Behaviour unchanged |
| [filegroup](https://www.php.net/manual/en/function.filegroup.php) | [https://3v4l.org/SZlZ9#veol](https://3v4l.org/SZlZ9#veol) | Behaviour unchanged |
| [fileinode](https://www.php.net/manual/en/function.fileinode.php) | [https://3v4l.org/N5O7I#veol](https://3v4l.org/N5O7I#veol) | Behaviour unchanged |
| [filemtime](https://www.php.net/manual/en/function.filemtime.php) | [https://3v4l.org/UZBYF#veol](https://3v4l.org/UZBYF#veol) | Behaviour unchanged |
| [fileowner](https://www.php.net/manual/en/function.fileowner.php) | [https://3v4l.org/ObITj#veol](https://3v4l.org/ObITj#veol) | Behaviour unchanged |
| [fileperms](https://www.php.net/manual/en/function.fileperms.php) | [https://3v4l.org/oOttL#veol](https://3v4l.org/oOttL#veol) | Behaviour unchanged |
| [filesize](https://www.php.net/manual/en/function.filesize.php) | [https://3v4l.org/3L2Qd#veol](https://3v4l.org/3L2Qd#veol) | Behaviour unchanged |
| [filetype](https://www.php.net/manual/en/function.filetype.php) | [https://3v4l.org/mPfKV#veol](https://3v4l.org/mPfKV#veol) | Behaviour unchanged |
| [fopen](https://www.php.net/manual/en/function.fopen.php) | [https://3v4l.org/hKUGL#veol](https://3v4l.org/hKUGL#veol) | Behaviour unchanged |
| [is_dir](https://www.php.net/manual/en/function.is-dir.php) | [https://3v4l.org/gQbqO#veol](https://3v4l.org/gQbqO#veol) | Behaviour unchanged |
| [is_executable](https://www.php.net/manual/en/function.is-executable.php) | [https://3v4l.org/WWppK1#veol](https://3v4l.org/WWppK1#veol) | Behaviour unchanged |
| [is_file](https://www.php.net/manual/en/function.is-file.php) | [https://3v4l.org/SerMJ#veol](https://3v4l.org/SerMJ#veol) | Behaviour unchanged |
| [is_link](https://www.php.net/manual/en/function.is-link.php) | [https://3v4l.org/X691s#veol](https://3v4l.org/X691s#veol) | Behaviour unchanged |
| [is_readable](https://www.php.net/manual/en/function.is-readable.php) | [https://3v4l.org/AlnoG#veol](https://3v4l.org/AlnoG#veol) | Behaviour unchanged |
| [is_writable](https://www.php.net/manual/en/function.is-writable.php) | [https://3v4l.org/5I6s3#veol](https://3v4l.org/5I6s3#veol) | Behaviour unchanged |
| [is_writeable](https://www.php.net/manual/en/function.is-writeable.php) | [https://3v4l.org/gp5Yl#veol](https://3v4l.org/gp5Yl#veol) | Behaviour unchanged |
| [lstat](https://www.php.net/manual/en/function.lstat.php) | [https://3v4l.org/4CcKp#veol](https://3v4l.org/4CcKp#veol) | Behaviour unchanged |

**Comment by d-alleyne** [comment]
@jrfnl I've observed that PHP 8.0 throws errors when there's a type mismatch in function arguments, for instance, passing an array where a string is expected. Should we classify these occurrences as behavioural changes?
